crates/oxide/src/glob.rs

```diff
@@ -1,11 +1,117 @@
-use fast_glob::glob_match;
+use crate::PublicSourceEntry;
 use fxhash::{FxHashMap, FxHashSet};
-use std::path::{Path, PathBuf};
-use tracing::event;
+use std::path::PathBuf;
+use tracing::{event, Level};
 
-use crate::GlobEntry;
+#[derive(Debug, Clone, PartialEq)]
+pub struct GlobEntry {
+    /// Base path of the glob
+    pub base: String,
 
-pub fn hoist_static_glob_parts(entries: &Vec<GlobEntry>) -> Vec<GlobEntry> {
+    /// Glob pattern
+    pub pattern: String,
+}
+
+/// Optimize the PublicSourceEntry by trying to move all the static parts of the pattern to the
+/// base of the PublicSourceEntry.
+///
+/// ```diff
+/// - { base: '/', pattern: 'src/**/*.html'}
+/// + { base: '/src', pattern: '**/*.html'}
+/// ```
+///
+/// A file stays in the `pattern` part, because the `base` should only be a directory.
+///
+/// ```diff
+/// - { base: '/', pattern: 'src/examples/index.html'}
+/// + { base: '/src/examples', pattern: 'index.html'}
+/// ```
+///
+/// A folder will be moved to the `base` part, and the `pattern` will be set to `**/*`.
+///
+/// ```diff
+/// - { base: '/', pattern: 'src/examples'}
+/// + { base: '/src/examples', pattern: '**/*'}
+/// ```
+///
+/// In addition, we will canonicalize the base path so we always work with the correctly resolved
+/// path.
+pub fn optimize_public_source_entry(source: &mut PublicSourceEntry) {
+    // Resolve base path immediately
+    let Ok(base) = dunce::canonicalize(&source.base) else {
+        event!(Level::ERROR, "Failed to resolve base: {:?}", source.base);
+        return;
+    };
+    source.base = base.to_string_lossy().to_string();
```

Comment:
This might be kinda strange to do but should we add a test for the non-utf-8 file path case? (idk if that's even feasible)